src/Nethermind/Nethermind.Evm/VirtualMachine.cs

```diff
@@ -1246,12 +1246,73 @@ static void UpdateCurrentState(EvmState state, int pc, long gas, int stackHead)
                     }
                 case Instruction.EXTCODESIZE:
                     {
+                        Metrics.ExtCodeSizeOpcode++;
                         long gasCost = spec.GetExtCodeCost();
                         if (!UpdateGas(gasCost, ref gasAvailable)) goto OutOfGas;
 
                         Address address = stack.PopAddress();
                         if (!ChargeAccountAccessGas(ref gasAvailable, vmState, address, spec)) goto OutOfGas;
 
+                        if (programCounter < code.Length)
+                        {
+                            bool optimizeAccess = false;
+                            Instruction nextInstruction = (Instruction)code[programCounter];
+                            if (nextInstruction == Instruction.ISZERO)
+                            {
+                                optimizeAccess = true;
+                                Metrics.ExtCodeSizeOptimizedIsZero++;
+                            }
+                            else if ((nextInstruction == Instruction.GT || nextInstruction == Instruction.EQ) &&
+                                    stack.PeekUInt256IsZero())
+                            {
+                                optimizeAccess = true;
+                                stack.PopLimbo();
+                                if (nextInstruction == Instruction.GT)
+                                {
+                                    Metrics.ExtCodeSizeOptimizedGT++;
+                                }
+                                else if (nextInstruction == Instruction.EQ)
+                                {
+                                    Metrics.ExtCodeSizeOptimizedEQ++;
+                                }
+                            }
+
+                            if (optimizeAccess)
+                            {
+                                // EXTCODESIZE ISZERO/GT/EQ peephole optimization.
+                                // In solidity 0.8.1+: `return account.code.length > 0;`
+                                // is is a common pattern to check if address is a contract
+                                // however we can just check the address's loaded CodeHash
+                                // to reduce storage access from trying to load the code
+                                if (traceOpcodes)
+                                {
+                                    EndInstructionTrace(gasAvailable, vmState.Memory?.Size ?? 0);
+                                    StartInstructionTrace(Instruction.ISZERO, vmState, gasAvailable, programCounter, in stack);
+                                }
+
+                                programCounter++;
+                                if (!UpdateGas(GasCostOf.VeryLow, ref gasAvailable)) goto OutOfGas;
```

Comment:
This is a silent check on the cost of gas of `Instruction.GT` and `Instruction.EQ`. Maybe worth to mention in a comment?

Reply:
Added comment